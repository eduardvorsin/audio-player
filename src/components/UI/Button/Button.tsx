import React from "react";
import { StyledButton } from "./StyledButton";

type ValidElements = React.ElementType;

type InferElement<E> =
	E extends keyof JSX.IntrinsicElements
	? JSX.IntrinsicElements[E] extends React.DetailedHTMLProps<React.AnchorHTMLAttributes<any>, infer Elem>
	? Elem
	: never
	: HTMLElement;

type ButtonOwnProps<E extends ValidElements = typeof defaultElement> = {
	children: React.ReactNode,
	as?: E,
	className?: string,
	startIcon?: React.ReactNode,
	endIcon?: React.ReactNode,
	primary?: boolean,
	withoutVisibleText?: boolean,
	onClick?: React.MouseEventHandler<InferElement<E>>,
	onKeyDown?: React.KeyboardEventHandler<InferElement<E>>,
}

const defaultElement = 'button';


type ButtonProps<E extends ValidElements> = ButtonOwnProps<E> & Omit<React.ComponentPropsWithoutRef<E>, keyof ButtonOwnProps<E>>;

export const Button = <E extends ValidElements = typeof defaultElement>({
	children = null,
	className = '',
	onClick = () => { },
	onKeyDown = () => { },
	startIcon = <></>,
	endIcon = <></>,
	primary = false,
	withoutVisibleText = false,
	as,
	...props
}: ButtonProps<E>) => {
	const Element: React.ElementType = as || defaultElement;

	const clickHandler: React.MouseEventHandler<InferElement<E>> = (e) => {
		onClick(e);
	}

	const keyDownHandler: React.KeyboardEventHandler<InferElement<E>> = (e) => {
		onKeyDown(e);
	}

	return (
		<StyledButton
			as={Element}
			className={className}
			onClick={clickHandler}
			onKeyDown={keyDownHandler}
			primary={primary}
			withoutVisibleText={withoutVisibleText}
			data-testid={Element}
			{...props}
		>
			{startIcon}
			{children}
			{endIcon}
		</StyledButton>
	)
};
