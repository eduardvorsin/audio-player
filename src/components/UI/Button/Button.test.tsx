import React from "react";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { Button } from "./Button";
import { ReactComponent as MockIcon } from '../../../assets/images/icons/play.svg';

describe('Button component tests', () => {

  test('renders correctly', () => {
    render(
      <Button>
        test text
      </Button>
    );

    expect(screen.getByRole<HTMLButtonElement>('button')).toBeInTheDocument();
  });

  test('when the button is clicked, the mock function is triggered', async () => {
    const user = userEvent.setup();
    const mockClickHandler = jest.fn<void, [React.MouseEvent<HTMLButtonElement>]>();
    render(
      <Button onClick={mockClickHandler}>
        test text
      </Button>
    );

    await user.click(screen.getByRole<HTMLButtonElement>('button'));

    expect(mockClickHandler).toBeCalledTimes(1);
  });

  test('when the button is pressed through the keyboard, the mock function is triggered', async () => {
    const user = userEvent.setup();
    const mockKeydownHandler = jest.fn<void, [React.KeyboardEvent<HTMLButtonElement>]>();
    render(
      <Button onKeyDown={mockKeydownHandler}>
        test text
      </Button>
    );

    await user.click(screen.getByRole<HTMLButtonElement>('button'));
    await user.keyboard('b');
    await user.keyboard('c');

    expect(mockKeydownHandler).toBeCalledTimes(2);
  });

  test('if the startIcon prop is passed, it renders the icon before the text', () => {
    render(
      <Button startIcon={<MockIcon title='play icon' />}>
        test text
      </Button>
    );

    expect(screen.getByTitle<HTMLElement>(/play icon/i)).toBeInTheDocument();
  });

  test('if the startIcon prop is passed, it renders the icon after the text', () => {
    render(
      <Button endIcon={<MockIcon title='play icon' />}>
        test text
      </Button>
    );

    expect(screen.getByTitle<HTMLElement>(/stop icon/i)).toBeInTheDocument();
  });

  test('if startIcon and endIcon prop are passed', () => {
    render(
      <Button
        endIcon={<MockIcon title='play icon' />}
        startIcon={<MockIcon title='stop icon' />}
      >
        test text
      </Button>
    );

    expect(screen.getByTitle<HTMLElement>(/play icon/i)).toBeInTheDocument();
    expect(screen.getByTitle<HTMLElement>(/stop icon/i)).toBeInTheDocument();
  });

  test('className is assigned correctly', () => {
    render(
      <Button
        className={'test'}
      >
        test text
      </Button>
    );

    expect(screen.getByRole<HTMLButtonElement>('button')).toHaveClass('test');
  });

  test('snapshot with default prop values', () => {
    render(
      <Button>
        test text
      </Button>
    );

    expect(screen.getByTestId<HTMLButtonElement>('button')).toMatchSnapshot();
  });

  test('snapshot with startIcon prop', () => {
    render(
      <Button
        startIcon={<MockIcon title='start icon' />}
      >
        test text
      </Button>
    );

    expect(screen.getByTestId<HTMLButtonElement>('button')).toMatchSnapshot();
  });


  test('snapshot with endIcon prop', () => {
    render(
      <Button
        endIcon={<MockIcon title='end icon' />}
      >
        test text
      </Button>
    );

    expect(screen.getByTestId<HTMLButtonElement>('button')).toMatchSnapshot();
  });

  test('snapshot with startIcon & endIcon props', () => {
    render(
      <Button
        startIcon={<MockIcon title='start icon' />}
        endIcon={<MockIcon title='end icon' />}
      >
        test text
      </Button>
    );

    expect(screen.getByTestId<HTMLButtonElement>('button')).toMatchSnapshot();
  });
});
