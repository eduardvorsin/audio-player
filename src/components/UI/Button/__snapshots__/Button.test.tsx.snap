// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component tests snapshot with as="a" prop 1`] = `
<a
  class="sc-bcXHqe dGjfYA"
  data-testid="button"
/>
`;

exports[`Button component tests snapshot with default prop values 1`] = `
<button
  class="sc-bcXHqe dGjfYA"
  data-testid="button"
/>
`;

exports[`Button component tests snapshot with endIcon prop 1`] = `
<button
  class="sc-bcXHqe dGjfYA"
  data-testid="button"
>
  <svg
    title="end icon"
  />
</button>
`;

exports[`Button component tests snapshot with startIcon & endIcon props 1`] = `
<button
  class="sc-bcXHqe dGjfYA"
  data-testid="button"
>
  <svg
    title="start icon"
  />
  <svg
    title="end icon"
  />
</button>
`;

exports[`Button component tests snapshot with startIcon prop 1`] = `
<button
  class="sc-bcXHqe dGjfYA"
  data-testid="button"
>
  <svg
    title="start icon"
  />
</button>
`;
